controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java

```diff
@@ -334,9 +334,10 @@ else if ( ! application.deploying().isPresent() && ! zone.environment().isManual
             Deployment previousDeployment = application.deployments().get(zone);
             Deployment newDeployment = previousDeployment;
             if (previousDeployment == null) {
-                newDeployment = new Deployment(zone, revision, version, clock.instant(), new HashMap<>(), new HashMap<>());
+                newDeployment = new Deployment(zone, revision, version, clock.instant());
```

Comment:
Could you encapsulate this stuff into a factory method on Deployment? 

I think all of this (line 334 to 340) should just be
`Deployment newDeployment = application.deployments().getOrDefault(new Deployment(zone)).with(revision).at(time);
`

Reply:
What should we default revision, version and deploy-time to in that scheme? And would that instance be valid (before adding revision, version and time)?

Reply:
Alternatively we could us a builder....

Reply:
I went with a middle ground here - which I think served the intent without considering the validity questions I had.